plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -443,6 +462,57 @@ private static <K, V> Map<K, V> getAll(Cache<K, AtomicReference<V>> cache, Itera
         return result.buildOrThrow();
     }
 
+    private static <K, V> Map<K, V> getAll(
+            Cache<K, AtomicReference<V>> cache,
+            Iterable<K> keys,
+            Function<Collection<K>, Map<K, V>> bulkLoader,
+            Predicate<V> isSufficient,
+            BiFunction<V, V, V> merger)
+    {
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        Map<K, AtomicReference<V>> toLoad = new HashMap<>();
+
+        keys.forEach(key -> {
+            // make sure the value holder is retrieved before the new values are loaded
+            // so that in case of invalidation we will not set the stale value
+            AtomicReference<V> currentValueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+            V currentValue = currentValueHolder.get();
+            if (currentValue != null && isSufficient.test(currentValue)) {
+                result.put(key, currentValue);
+            }
+            else {
+                toLoad.put(key, currentValueHolder);
+            }
+        });
+
+        if (toLoad.isEmpty()) {
+            return result.buildOrThrow();
+        }
+
+        Map<K, V> newEntries = bulkLoader.apply(toLoad.keySet());
+        newEntries.forEach((key, newValue) -> {
+            AtomicReference<V> valueHolder = toLoad.get(key);
+            result.put(key, merge(merger, key, valueHolder, valueHolder.get(), newValue));
```

Comment:
why provide both atomic reference and its current value to `merge` (aka oldValue)?
the value ultimately stored by `merge` should not depend on some  value retrieved in the past


```java
newEntries.forEach((key, newValue) -> {
            if (newValue == null) {
                throw new InvalidCacheLoadException("Failed to return a value for " + key);
            }
            AtomicReference<V> valueHolder = toLoad.get(key);
            verifyNotNull(valueHolder, "Unexpected value for key %s that was not asked for", key);
            V merged = valueHolder.accumulateAndGet(newValue, (current, added) -> current == null ? added : merger.apply(current, added));
            result.put(key, merged);
        });
```

Reply:
merge is also used in the `getIncremental` and the old value is needed there to make sure we have all columns covered even if the invalidation happened

Reply:
the invalidation doesn't erase the valueholder.
the value held by the value holder will only "grow"

Reply:
right, I missed that that "valueHolder" will never shrink. merge method dropped